crates/bevy_tasks/src/task_pool.rs

```diff
@@ -105,56 +131,83 @@ impl TaskPool {
-        stack_size: Option<usize>,
-        thread_name: Option<&str>,
-    ) -> Self {
+    fn new_internal(builder: TaskPoolBuilder) -> Self {
         let (shutdown_tx, shutdown_rx) = async_channel::unbounded::<()>();
 
-        let executor = Arc::new(async_executor::Executor::new());
-
-        let num_threads = num_threads.unwrap_or_else(num_cpus::get);
-
-        let threads = (0..num_threads)
-            .map(|i| {
-                let ex = Arc::clone(&executor);
-                let shutdown_rx = shutdown_rx.clone();
-
-                // miri does not support setting thread names
-                // TODO: change back when https://github.com/rust-lang/miri/issues/1717 is fixed
-                #[cfg(not(miri))]
-                let mut thread_builder = {
-                    let thread_name = if let Some(thread_name) = thread_name {
-                        format!("{} ({})", thread_name, i)
-                    } else {
-                        format!("TaskPool ({})", i)
-                    };
-                    thread::Builder::new().name(thread_name)
-                };
-                #[cfg(miri)]
-                let mut thread_builder = {
-                    let _ = i;
-                    let _ = thread_name;
-                    thread::Builder::new()
-                };
-
-                if let Some(stack_size) = stack_size {
-                    thread_builder = thread_builder.stack_size(stack_size);
-                }
-
-                thread_builder
-                    .spawn(move || {
-                        let shutdown_future = ex.run(shutdown_rx.recv());
-                        // Use unwrap_err because we expect a Closed error
-                        future::block_on(shutdown_future).unwrap_err();
-                    })
-                    .expect("Failed to spawn thread.")
-            })
-            .collect();
+        let compute_executor = Arc::new(async_executor::Executor::new());
+        let async_compute_executor = Arc::new(async_executor::Executor::new());
+        let io_executor = Arc::new(async_executor::Executor::new());
+
+        let compute_threads = builder.compute_threads.unwrap_or_else(num_cpus::get);
+        let io_threads = builder.io_threads.unwrap_or(0);
+        let async_compute_threads = builder.async_compute_threads.unwrap_or(0);
+
+        let mut threads = Vec::with_capacity(compute_threads + io_threads + async_compute_threads);
+        threads.extend((0..compute_threads).map(|i| {
+            let thread_builder = make_thread_builder(
+                builder.thread_name.as_deref(),
+                "Compute",
+                i,
+                builder.stack_size,
+            );
+            let compute = Arc::clone(&compute_executor);
+            let shutdown_rx = shutdown_rx.clone();
+
+            thread_builder
+                .spawn(move || {
+                    let future = compute.run(shutdown_rx.recv()).or(shutdown_rx.recv());
```

Comment:
why does this need to have shutdown_rx.recv() twice?

Reply:
Whoops. Nice catch.